R/aggregate.R

```diff
@@ -180,43 +143,380 @@ agg_climate_data <- function(year, data_raster, climate_var, daily_agg, trans =
+    dt <- as.data.table.raster(r[[x]], xy=TRUE)
+
+    # Set column names with months
+    new_names <- c('x', 'y', layer_names)
+    data.table::setnames(dt, new_names)
+
+    # Change from wide to long format
+    dt = data.table::melt(dt, id.vars = c("x", "y"))
+
+    # Update variable names
+    var_names <- c('date', list_names[x])
+    data.table::setnames(dt, old=c('variable', 'value'), new=var_names)
+  }
+
+  # Make each raster layer a data.table
+  list_dt <- lapply(1:(list_length + 1), create_dt)
+
+  # Merge all data tables together
+  clim_dt <- list_dt[[1]]
+
+  i <- 2
+  while(i <= list_length + 1){
+    dt_m <- list_dt[[i]]
+    clim_dt <- merge(clim_dt, dt_m, by=c('x', 'y', 'date'))
+    i <- i + 1
+  }
+
+
+
+  # Aggregate by polygon
+  sum_by_poly <- after_trans(clim_dt, weights_dt, list_names, second_weights, agg_to)
+
+  return(sum_by_poly)
+}
+
+
+
+#' @export
+staggregate_bin <- function(data_raster, climate_var, daily_agg, num_bins = 30, binwidth = NULL, center_on = NULL, start_on = NULL, end_on = NULL, max = NULL, min = NULL, geoweights_table, second_weights, agg_to = "month"){
+  # Get climate data as a data table and aggregate to daily values
+  setup_list <- before_trans(data_raster, climate_var, daily_agg, geoweights_table, second_weights)
+  clim_daily <- setup_list[[1]]
+  weights_dt <- setup_list[[2]]
+  layer_names <- setup_list[[3]]
+
+
+  clim_daily_table <- raster::values(clim_daily)
+
+  if(is.null(max)){
+    max <- max(clim_daily_table) + .01
+  }
+  if(is.null(min)){
+    min <- min(clim_daily_table) - .01
+  }
+
+  # Write message that binwidth overrides num_bins
+  if(!is.null(binwidth)){
+    message(crayon::yellow("Binwidth argument supplied will override num_bins"))
+    num_bins <- ceiling((max - min) / binwidth)
+  }
+
+  # If value not supplied to binwidth, calculate from num_bins
+  if(is.null(binwidth)){
+    binwidth <- (max - min) / num_bins
+  }
+
+  # Stop with message that only one of center_on, start_on, and end_on can be chosen if otherwise
+  if((!is.null(center_on) & !is.null(start_on)) |
+     (!is.null(center_on) & !is.null(end_on)) |
+     (!is.null(start_on) & !is.null(end_on))){
+    stop(crayon::red("Too many bin placement arguments specified. Please specify a value for only one variable out of center_on, start_on, or end_on"))
+  }
+
+  if(is.null(center_on) & is.null(start_on) & is.null(end_on)){
+    message(crayon::yellow("No bin placement argument specified. Drawing bins from min value."))
+    start_on <- min
+  }
+
+  # Put start_on and end_on in terms of center_on
+  if(!is.null(start_on)){
+    center_on <- start_on + (binwidth / 2)
+  }
+
+  if(!is.null(end_on)){
+    center_on <- end_on - (binwidth / 2)
+  }
+
+
+
+  # Check to see that center_on falls within the range of data, move it if not
+  if(max - min < binwidth){
+    stop(crayon::red("The binwidth is larger than the range of data. Please specify a different binwidth and rerun"))
+  }
+
+  if(center_on > max){
```

Comment:
i don't think we want this conditional and/or warning, as mentioned above with just one year of data being loaded at a time. totally ok to just have all zeros for that bin.

Reply:
The reason this stop is called is because in adjusting the `center_on` value, it is possible that a `binwidth` bigger than `max - min` will cause us to jump completely over the range. If for example `min = 0, max = 4, binwidth = 6, center_on = 5` we'd adjust `center_on` down to -1, and then because it's lower than `min`, back up to 5, at which point it will try to move on to drawing the bins.